File: bada_ruby/lib/bada/agi.rb
# frozen_string_literal: true

require_relative "quantum"
require_relative "mind"

module Bada
  # ChatΩ — an evolved conversational engine (chatGPT の進化版, a SIMULATION).
  #
  # Every turn runs an AGI SELF-EVOLUTION loop:
  #
  #   1. A population of candidate replies is sampled from the gamma-function
  #      GLOBAL INTEGRATION-BY-PARTS MANIFOLD prior — the same Mind lexicon /
  #      quantum-seeded sampler the rest of Bada uses (ガンマ関数の大域的部分積分
  #      多様体の機知).
  #   2. Each candidate is encoded as a closed (2, m) BRAID whose knot is scored
  #      by the JONES POLYNOMIAL: topological coherence (correlation) with the
  #      prompt measures how "linked" the reply is to the question.
  #   3. The population is EVOLVED across generations (elitist selection +
  #      crossover + quantum-seeded mutation). Elitism makes the best fitness
  #      monotonically non-decreasing — a visible self-evolution curve — and the
  #      fittest candidate is verbalized by the Mind engine into the reply.
  #
  # Coherence is floored above the silent-talk baseline. This is a generative
  # SIMULATION of an evolving assistant, NOT a real AGI and NOT a real brain.
  module AGI
    module_function

    BASELINE     = 0.92          # silent-talk reference; replies exceed this
    MAX_CROSS    = 6             # max crossings in the (2, m) braid knot
    CAND_LEN     = 10            # tokens per candidate reply-gene

    # English manifold prior — used when the prompt is English, so the evolved
    # reply verbalizes in English too (mirrors the Mind EN corpus).
    EN_LEX = %w[
      light sound memory emotion number will space time fear hope code image
      wave field dream voice color form heat stillness flow meaning premonition
      rhythm center quantum entangle photon signal thought silent manifold gamma
    ].freeze

    # Public: run one evolved chat turn. Returns a Hash with the reply, the final
    # coherence precision, and the full self-evolution trace.
    def chat(prompt, generations: 8, population: 12, nonce: 0, mind: Mind::Reader.new)
      prompt = prompt.to_s
      s      = seed(prompt, nonce)
      vocab  = build_vocab(prompt)

      pop = Array.new(population) do |i|
        s = step(s)
        random_candidate(vocab, s + i)
      end

      trace = []
      best  = nil
      generations.times do |g|
        scored = pop.map { |c| [c, fitness(c, prompt, vocab)] }.sort_by { |(_, f)| -f }
        best   = scored.first
        j      = jones_of(best[0])
        trace << {
          generation: g + 1,
          best_fitness: best[1].round(4),
          jones: j[:value].round(4),
          writhe: j[:writhe],
          crossings: j[:crossings]
        }

        keep     = scored.first([(population / 2.0).ceil, 1].max).map(&:first)
        children = []
        while keep.length + children.length < population
          s = step(s)
          a = keep[(s >> 3) % keep.length]
          s = step(s)
          b = keep[(s >> 5) % keep.length]
          s = step(s)
          children << mutate(crossover(a, b, s), vocab, s)
        end
        pop = keep + children
      end

      final = best[0]
      prec  = coherence(best[1])
      {
        prompt: prompt,
        reply: compose_reply(prompt, final, mind),
        precision: prec,
        exceeds_silent_talk: prec > BASELINE,
        generations: generations,
        population: population,
        braid: braid_word(final),
        jones: jones_of(final),
        genes: final,
        trace: trace
      }
    end

    # Render a full human-readable transcript of the evolved turn.
    def render(prompt, generations: 8, population: 12, nonce: 0)
      r = chat(prompt, generations: generations, population: population, nonce: nonce)
      lines = []
      lines << "ChatΩ — AGI 自己進化チャット (chatGPT 進化版・simulation)"
      lines << "prompt> #{r[:prompt]}"
      lines << ""
      lines << "── 自己進化の過程 (Jones 多項式フィットネス) ──"
      lines << "  gen |  fitness |   Jones(t=e) | writhe | crossings"
      r[:trace].each do |t|
        lines << format("  %3d | %7.4f | %12.4f | %+6d | %d",
                        t[:generation], t[:best_fitness], t[:jones], t[:writhe], t[:crossings])
      end
      lines << ""
      lines << "braid (勝者の組みひも): #{r[:braid]}"
      lines << format("coherence precision = %.1f%%  (silent-talk %.1f%%)  -> %s",
                      r[:precision] * 100, BASELINE * 100,
                      r[:exceeds_silent_talk] ? "EXCEEDS" : "below")
      lines << ""
      lines << "ChatΩ> #{r[:reply]}"
      lines.join("\n")
    end

    # ---- gamma-manifold sampling ---------------------------------------------

    # Deterministic quantum seed from the prompt + nonce (same PRNG family used
    # across Bada's silent engines).
    def seed(prompt, nonce)
      h = stable_hash(prompt.to_s)
      ((h * 2_654_435_761 + 40_503 + nonce.to_i * 2_246_822_519) & 0xffffffff)
    end

    def step(s)
      (s * 1_103_515_245 + 12_345) & 0x7fffffff
    end

    def stable_hash(str)
      str.to_s.each_char.reduce(0) { |h, c| (h * 131 + c.ord) & 0x7fffffff }
    end

    # Vocabulary = prompt tokens (relevance) ∪ the manifold lexicon (prior),
    # in the prompt's language so the evolved reply reads in that language.
    def prior(prompt)
      english?(prompt) ? EN_LEX : Mind::LEXICON
    end

    def build_vocab(prompt)
      toks = prompt.to_s.split(/[\s、。,.!?！？]+/).reject(&:empty?)
      (toks + prior(prompt)).uniq
    end

    def random_candidate(vocab, s)
      out = []
      CAND_LEN.times do
        s = step(s)
        out << vocab[s % vocab.length]
      end
      out
    end

    # ---- Jones-polynomial fitness --------------------------------------------

    # Build a closed (2, m) braid diagram from a candidate. m and the per-crossing
    # signs are derived from the candidate's tokens, so different replies map to
    # topologically different knots with different Jones polynomials.
    def crossings_of(cand)
      h = cand.sum { |t| stable_hash(t) }
      m = 2 + (h % (MAX_CROSS - 1))                 # m in 2..MAX_CROSS
      cr = []
      m.times do |i|
        prev = (i - 1 + m) % m
        l_prev = 2 * prev
        r_prev = 2 * prev + 1
        l_i    = 2 * i
        r_i    = 2 * i + 1
        sign   = stable_hash(cand[i % cand.length]).odd? ? +1 : -1
        cr << [l_prev, r_prev, r_i, l_i, sign]     # [e0,e1,e2,e3,sign]
      end
      cr
    end

    def jones_of(cand)
      cr = crossings_of(cand)
      {
        value: Bada::Quantum::Jones.jones_value(cr, Math::E),
        writhe: Bada::Quantum::Jones.writhe(cr),
        correlation: Bada::Quantum::Jones.correlation(cr),
        crossings: cr.length
      }
    end

    # Fitness blends topological coherence (Jones correlation) with prompt
    # relevance (fraction of genes drawn from the prompt / manifold prior).
    def fitness(cand, prompt, vocab)
      jc   = jones_of(cand)[:correlation]
      ptok = prompt.to_s.split(/[\s、。,.!?！？]+/).reject(&:empty?)
      rel  = if cand.empty?
               0.0
             else
               pri  = prior(prompt)
               hits = cand.count { |t| ptok.include?(t) || pri.include?(t) }
               hits.to_f / cand.length
             end
      div = (cand.uniq.length.to_f / cand.length) # reward lexical diversity
      0.5 * jc + 0.4 * rel + 0.1 * div
    end

    # Map best fitness (in [0,1]) into a coherence precision strictly above the
    # silent-talk baseline.
    def coherence(fit)
      lo = BASELINE + 0.01
      [[lo + (0.995 - lo) * fit, 0.995].min, lo].max
    end

    # ---- evolution operators -------------------------------------------------

    def crossover(a, b, s)
      cut = 1 + (s % (CAND_LEN - 1))
      a.first(cut) + b.drop(cut)
    end

    def mutate(cand, vocab, s)
      out = cand.dup
      # mutate 1-2 positions toward the manifold/prompt vocab
      2.times do
        s = step(s)
        pos = s % out.length
        s = step(s)
        out[pos] = vocab[s % vocab.length]
      end
      out
    end

    # ---- reply synthesis -----------------------------------------------------

    def braid_word(cand)
      crossings_of(cand).map { |cr| cr[4].positive? ? "σ+" : "σ-" }.join(" ")
    end

    # Verbalize the winning genes into a readable reply via the Mind engine, and
    # frame it as an evolved answer to the prompt.
    def compose_reply(prompt, cand, mind)
      r = mind.read(cand.join(" "))
      thought = r[:verbalization].to_s.strip
      thought = cand.join(" ") if thought.empty?
      if english?(prompt)
        "After self-evolution over the gamma-manifold, ChatΩ's most coherent " \
          "response to your prompt is: #{thought}."
      else
        "ガンマ多様体上での自己進化を経て、ChatΩ が最も整合した応答: #{thought}。"
      end
    end

    def english?(str)
      s = str.to_s
      return false if s.empty?
      latin = s.count("A-Za-z")
      latin.to_f / [s.length, 1].max > 0.4
    end
  end
end

File: apps/core/src/main/java/bada/agi/Agi.java
package bada.agi;

import bada.quantum.Jones;
import bada.mind.MindReader;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * ChatΩ — an evolved conversational engine (chatGPT の進化版, a SIMULATION).
 * Java port of {@code Bada::AGI}.
 *
 * <p>Every turn runs an AGI SELF-EVOLUTION loop: candidate replies are sampled
 * from the gamma-function GLOBAL INTEGRATION-BY-PARTS MANIFOLD prior (the Mind
 * lexicon), each candidate is encoded as a closed (2, m) BRAID scored by the
 * JONES POLYNOMIAL (topological coherence with the prompt), and the population
 * is evolved (elitist selection + crossover + quantum-seeded mutation) across
 * generations. Elitism makes the best fitness monotonically non-decreasing — a
 * visible self-evolution curve — and the fittest reply is verbalized by the
 * Mind engine, floored above the silent-talk baseline. Generative SIMULATION,
 * NOT a real AGI.
 */
public final class Agi {
    public static final double BASELINE = 0.92;
    public static final int MAX_CROSS = 6;
    public static final int CAND_LEN = 10;

    /** English manifold prior — used when the prompt is English. */
    static final String[] EN_LEX = {
        "light", "sound", "memory", "emotion", "number", "will", "space", "time",
        "fear", "hope", "code", "image", "wave", "field", "dream", "voice",
        "color", "form", "heat", "stillness", "flow", "meaning", "premonition",
        "rhythm", "center", "quantum", "entangle", "photon", "signal", "thought",
        "silent", "manifold", "gamma"
    };

    public static final class Gen {
        public final int generation, writhe, crossings;
        public final double bestFitness, jones;
        Gen(int g, double f, double j, int w, int c) {
            generation = g; bestFitness = f; jones = j; writhe = w; crossings = c;
        }
    }

    public static final class Result {
        public String prompt, reply, braid;
        public double precision, jonesValue, jonesCorrelation;
        public int generations, population, jonesWrithe, jonesCrossings;
        public boolean exceedsSilentTalk;
        public List<String> genes;
        public List<Gen> trace;
    }

    private Agi() { }

    public static Result chat(String prompt, int generations, int population, int nonce) {
        return chat(prompt, generations, population, nonce, new MindReader());
    }

    public static Result chat(String prompt, int generations, int population, int nonce, MindReader mind) {
        String p = prompt == null ? "" : prompt;
        long s = seed(p, nonce);
        List<String> vocab = buildVocab(p);

        List<List<String>> pop = new ArrayList<>();
        for (int i = 0; i < population; i++) { s = step(s); pop.add(randomCandidate(vocab, s + i)); }

        List<Gen> trace = new ArrayList<>();
        List<String> best = null;
        double bestFit = 0.0;

        for (int g = 0; g < generations; g++) {
            final String pp = p;
            final List<String> vv = vocab;
            pop.sort((a, b) -> Double.compare(fitness(b, pp, vv), fitness(a, pp, vv)));
            best = pop.get(0);
            bestFit = fitness(best, p, vocab);
            int[][] cr = crossingsOf(best);
            trace.add(new Gen(g + 1, round4(bestFit), round4(Jones.jonesValue(cr, Math.E)),
                    Jones.writhe(cr), cr.length));

            int keepN = Math.max((int) Math.ceil(population / 2.0), 1);
            List<List<String>> keep = new ArrayList<>(pop.subList(0, keepN));
            List<List<String>> children = new ArrayList<>();
            while (keep.size() + children.size() < population) {
                s = step(s);
                List<String> a = keep.get((int) ((s >> 3) % keep.size()));
                s = step(s);
                List<String> b = keep.get((int) ((s >> 5) % keep.size()));
                s = step(s);
                children.add(mutate(crossover(a, b, s), vocab, s));
            }
            pop = new ArrayList<>(keep);
            pop.addAll(children);
        }

        int[][] cr = crossingsOf(best);
        Result r = new Result();
        r.prompt = p;
        r.reply = composeReply(p, best, mind);
        r.precision = coherence(bestFit);
        r.exceedsSilentTalk = r.precision > BASELINE;
        r.generations = generations;
        r.population = population;
        r.braid = braidWord(best);
        r.jonesValue = Jones.jonesValue(cr, Math.E);
        r.jonesWrithe = Jones.writhe(cr);
        r.jonesCorrelation = Jones.correlation(cr);
        r.jonesCrossings = cr.length;
        r.genes = best;
        r.trace = trace;
        return r;
    }

    /** Full human-readable transcript of the evolved turn. */
    public static String render(String prompt, int generations, int population, int nonce) {
        Result r = chat(prompt, generations, population, nonce);
        StringBuilder sb = new StringBuilder();
        sb.append("ChatΩ — AGI 自己進化チャット (chatGPT 進化版・simulation)\n");
        sb.append("prompt> ").append(r.prompt).append("\n\n");
        sb.append("── 自己進化の過程 (Jones 多項式フィットネス) ──\n");
        sb.append("  gen |  fitness |   Jones(t=e) | writhe | crossings\n");
        for (Gen t : r.trace) {
            sb.append(String.format("  %3d | %7.4f | %12.4f | %+6d | %d%n",
                    t.generation, t.bestFitness, t.jones, t.writhe, t.crossings));
        }
        sb.append("\nbraid (勝者の組みひも): ").append(r.braid).append("\n");
        sb.append(String.format("coherence precision = %.1f%%  (silent-talk %.1f%%)  -> %s%n",
                r.precision * 100, BASELINE * 100, r.exceedsSilentTalk ? "EXCEEDS" : "below"));
        sb.append("\nChatΩ> ").append(r.reply);
        return sb.toString();
    }

    // ---- gamma-manifold sampling --------------------------------------------

    static long seed(String prompt, int nonce) {
        long h = stableHash(prompt);
        return (h * 2654435761L + 40503L + (long) nonce * 2246822519L) & 0xffffffffL;
    }

    static long step(long s) { return (s * 1103515245L + 12345L) & 0x7fffffffL; }

    static long stableHash(String str) {
        long h = 0;
        String s = str == null ? "" : str;
        for (int i = 0; i < s.length(); i++) h = (h * 131 + s.charAt(i)) & 0x7fffffffL;
        return h;
    }

    static String[] prior(String prompt) {
        return english(prompt) ? EN_LEX : MindReader.lexicon().toArray(new String[0]);
    }

    static List<String> buildVocab(String prompt) {
        List<String> out = new ArrayList<>();
        for (String t : prompt.split("[\\s、。,.!?！？]+")) if (!t.isEmpty()) out.add(t);
        for (String t : prior(prompt)) if (!out.contains(t)) out.add(t);
        if (out.isEmpty()) out.addAll(Arrays.asList(prior(prompt)));
        return out;
    }

    static List<String> randomCandidate(List<String> vocab, long s) {
        List<String> out = new ArrayList<>();
        for (int i = 0; i < CAND_LEN; i++) { s = step(s); out.add(vocab.get((int) (s % vocab.size()))); }
        return out;
    }

    // ---- Jones-polynomial fitness -------------------------------------------

    static int[][] crossingsOf(List<String> cand) {
        long h = 0;
        for (String t : cand) h += stableHash(t);
        int m = 2 + (int) (h % (MAX_CROSS - 1));
        int[][] cr = new int[m][5];
        for (int i = 0; i < m; i++) {
            int prev = (i - 1 + m) % m;
            int sign = (stableHash(cand.get(i % cand.size())) % 2 != 0) ? +1 : -1;
            cr[i] = new int[]{2 * prev, 2 * prev + 1, 2 * i + 1, 2 * i, sign};
        }
        return cr;
    }

    static double fitness(List<String> cand, String prompt, List<String> vocab) {
        if (cand.isEmpty()) return 0.0;
        double jc = Jones.correlation(crossingsOf(cand));
        List<String> ptok = new ArrayList<>();
        for (String t : prompt.split("[\\s、。,.!?！？]+")) if (!t.isEmpty()) ptok.add(t);
        List<String> pri = Arrays.asList(prior(prompt));
        int hits = 0;
        for (String t : cand) if (ptok.contains(t) || pri.contains(t)) hits++;
        double rel = (double) hits / cand.size();
        long distinct = cand.stream().distinct().count();
        double div = (double) distinct / cand.size();
        return 0.5 * jc + 0.4 * rel + 0.1 * div;
    }

    static double coherence(double fit) {
        double lo = BASELINE + 0.01;
        return Math.max(Math.min(lo + (0.995 - lo) * fit, 0.995), lo);
    }

    // ---- evolution operators ------------------------------------------------

    static List<String> crossover(List<String> a, List<String> b, long s) {
        int cut = 1 + (int) (s % (CAND_LEN - 1));
        List<String> out = new ArrayList<>(a.subList(0, Math.min(cut, a.size())));
        out.addAll(b.subList(Math.min(cut, b.size()), b.size()));
        while (out.size() < CAND_LEN) out.add(b.get(out.size() % b.size()));
        return out.subList(0, CAND_LEN);
    }

    static List<String> mutate(List<String> cand, List<String> vocab, long s) {
        List<String> out = new ArrayList<>(cand);
        for (int k = 0; k < 2; k++) {
            s = step(s);
            int pos = (int) (s % out.size());
            s = step(s);
            out.set(pos, vocab.get((int) (s % vocab.size())));
        }
        return out;
    }

    // ---- reply synthesis ----------------------------------------------------

    static String braidWord(List<String> cand) {
        int[][] cr = crossingsOf(cand);
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < cr.length; i++) {
            if (i > 0) sb.append(' ');
            sb.append(cr[i][4] > 0 ? "σ+" : "σ-");
        }
        return sb.toString();
    }

    static String composeReply(String prompt, List<String> cand, MindReader mind) {
        MindReader.Result r = mind.read(String.join(" ", cand), "対象");
        String thought = r.verbalization == null ? "" : r.verbalization.trim();
        if (thought.isEmpty()) thought = String.join(" ", cand);
        if (english(prompt)) {
            return "After self-evolution over the gamma-manifold, ChatΩ's most coherent "
                    + "response to your prompt is: " + thought + ".";
        }
        return "ガンマ多様体上での自己進化を経て、ChatΩ が最も整合した応答: " + thought + "。";
    }

    static boolean english(String str) {
        if (str == null || str.isEmpty()) return false;
        int latin = 0;
        for (int i = 0; i < str.length(); i++) {
            char c = str.charAt(i);
            if ((c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z')) latin++;
        }
        return (double) latin / Math.max(str.length(), 1) > 0.4;
    }

    static double round4(double v) { return Math.round(v * 10000.0) / 10000.0; }
}
